internal/controllers/machinedeployment/machinedeployment_rollout_rollingupdate.go

```diff
@@ -292,6 +357,152 @@ func (p *rolloutPlanner) scaleDownOldMSs(ctx context.Context, totalScaleDownCoun
+		// Check if the MachineSet can update in place; if not, move to the next MachineSet.
+		// TODO(in-place): replace with a call to the hook
+		canUpdateMachineSetInPlaceFunc := func(_ *clusterv1.MachineSet) bool { return false }
+		if p.overrideCanUpdateMachineSetInPlace != nil {
+			canUpdateMachineSetInPlaceFunc = p.overrideCanUpdateMachineSetInPlace
+		}
+
+		canUpdateInPlace := canUpdateMachineSetInPlaceFunc(oldMS)
+		log.V(5).Info(fmt.Sprintf("CanUpdate in-place decision for MachineSet %s: %t", oldMS.Name, canUpdateInPlace), "MachineSet", klog.KObj(oldMS))
+
+		if !canUpdateInPlace {
+			continue
+		}
+
+		// Set the annotation informing the oldMS that it must move machines to the newMS instead of deleting them.
+		// Note: After a machine is moved from oldMS to newMS, the newMS will take care of the in-place upgrade process.
+		// Note: Cleanup of the MachineSetMoveMachinesToAnnotationName annotation will happen automatically as soon as the rollout planner stops
+		// to set it, because this annotation is not part of the output of computeDesiredMS
+		// (same applies to newMS, so annotation will always be removed from newMS).
+		if oldMS.Annotations == nil {
+			oldMS.Annotations = map[string]string{}
+		}
+		oldMS.Annotations[clusterv1.MachineSetMoveMachinesToAnnotationName] = p.newMS.Name
+		inPlaceUpdateCandidates.Insert(oldMS.Name)
+	}
+
+	// If there are no inPlaceUpdateCandidates, nothing left to do.
+	if inPlaceUpdateCandidates.Len() <= 0 {
+		return nil
+	}
+
+	// Set the annotation informing the newMS that it will receive replicas moved from oldMS selected as in-place candidates.
+	// Note: there is a two-ways check before the move operation:
+	// 	"oldMS must have: move to newMS" and "newMS must have: accept replicas from oldMS"
+	// Note: Cleanup of the MachineSetReceiveMachinesFromAnnotationName annotation will happen automatically as soon as the rollout planner stops
+	// to set it, because this annotation is not part of the output of computeDesiredMS
+	// (same applies to oldMS, so annotation will always be removed from oldMS).
+	if p.newMS.Annotations == nil {
+		p.newMS.Annotations = map[string]string{}
+	}
+	p.newMS.Annotations[clusterv1.MachineSetReceiveMachinesFromAnnotationName] = sortAndJoin(inPlaceUpdateCandidates.UnsortedList())
+
+	// If the newMS is not scaling up, nothing left to do.
+	if scaleIntent, ok := p.scaleIntents[p.newMS.Name]; !ok || scaleIntent < ptr.Deref(p.newMS.Spec.Replicas, 0) {
+		return nil
+	}
+
+	// If the newMS is scaling up while there are still in place updates to be performed,
+	// check if the current scale up intent is using maxSurge.
+	scaleUpCount := p.scaleIntents[p.newMS.Name] - ptr.Deref(p.newMS.Spec.Replicas, 0)
+	scaleUpCountWithoutMaxSurge := max(ptr.Deref(p.md.Spec.Replicas, 0)-mdutil.TotalMachineSetsReplicaSum(allMSs), 0)
+	if scaleUpCount <= scaleUpCountWithoutMaxSurge {
+		// current scale up intent for the newMS is not using maxSurge, no need to revisit it.
+		return nil
+	}
```

Comment:
```suggestion
	// Check if the newMS is using maxSurge to scale up.
	scaleUpCount := p.scaleIntents[p.newMS.Name] - ptr.Deref(p.newMS.Spec.Replicas, 0)
	scaleUpCountWithoutMaxSurge := max(scaleUpCount-mdutil.MaxSurge(*p.md), 0)
	maxSurgeUsed := scaleUpCount - scaleUpCountWithoutMaxSurge
	if maxSurgeUsed <= 0 {
		// current scale up intent for the newMS is not using maxSurge, no need to revisit it.
		return nil
	}
```

What do you think about this?